packages/react-i18n/src/index.tsx

```diff
@@ -88,15 +82,27 @@ type Optionalize< T extends K, K > = Omit< T, keyof K >;
  * export default withI18n( MyComponent );
  */
 export const withI18n = < T extends I18nProps = I18nProps >(
-	WrappedComponent: React.ComponentType< T >
-): React.FunctionComponent< Optionalize< T, I18nProps > > => props => {
-	const i18n = useI18n();
-	return (
-		<WrappedComponent
-			{ ...i18n }
-			// Required cast `props as T`
-			// See https://github.com/Microsoft/TypeScript/issues/28938
-			{ ...( props as T ) }
-		/>
+	InnerComponent: React.ComponentType< T >
+): React.FunctionComponent< Optionalize< T, I18nProps > > => {
+	const WrappedComponent: React.FunctionComponent< Optionalize< T, I18nProps > > = props => (
+		<I18nContext.Consumer>
+			{ i18n => {
+				if ( i18n === undefined ) {
+					throw new Error( 'withI18n must be descendent of <I18nProvider />' );
+				}
+				return (
+					<InnerComponent
+						{ ...i18n }
+						// Required cast `props as T`
+						// See https://github.com/Microsoft/TypeScript/issues/28938
+						{ ...( props as T ) }
+					/>
+				);
+			} }
+		</I18nContext.Consumer>
 	);
+	WrappedComponent.displayName = `withI18n( ${ WrappedComponent.displayName ||
+		WrappedComponent.name ||
+		'Component' } )`;
```

Comment:
Can you use `createHigherOrderComponent` from `@wordpress/compose`? Or is that not TypeScript friendly?

Reply:
Done, the types work nicely 👍